### PR TITLE
Compare datasets - Integrate Quisby into Pbench Server API

### PR DESCRIPTION
**Pull request by siddardh-ra**
Implementation of API to compare multiple datasets using quisby processing

`GET /api/v1/compare?datasets=d1,d2,d3`

Currently, the pquisby package supports only the uperf benchmark

### CODE REVIEW

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:66]: "requested access"? You lost a space here. (Or, even worse, it was missing from your source?)

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:82]: Instead, declare the `datasets` schema parameter as `element_type=ParamType.DATASET,` and the normalization will be done automatically as part of list processing!

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:50]: This is interesting. In theory maybe `ApiAuthorizationType.DATASET` could be made to work here by recognizing that it's a list and validating *all* of them; but I'm pretty sure it won't do that.

Without that, however, you will have to manually authenticate against each dataset in the list to ensure that the user has READ access. Something like

```python
        self._check_authorization(
            ApiAuthorization(
                ApiAuthorizationType.USER_ACCESS,
                OperationCode.READ,
                str(dataset.owner_id),
                dataset.access,
            )
        )
```

I think the code as written now will allow comparing anyone's private datasets.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:91]: I don't know why I didn't think of it in the other PR, but this is cumbersome. Instead, try just `benchmark = Metadata.getvalue(dataset, "dataset.metalog.pbench.script")` since you're only getting *one*.

The helper is designed for getting a whole array of metadata keys at once, and you're only getting one.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:85]: `benchmark_check = set()` will create a set in the first place. You don't need to accumulate the entire list first and then convert! (Use `benchmark_check.add()` instead of `benchmark_check.append()`.)

Also, at the end, you could consider adding the set (or `",".join(set)`) to the `APIAbort` message giving the list of benchmark types they specified...

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:114]: This isn't wrong, really, but it's not going to accommodate roles. Right now, that means an `ADMIN` can't compare two arbitrary datasets. I'm not sure I really care about that, but as (if) we add support for groups and more general roles, it'll be more important to use the standard authorization check, as I suggested above.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:122]: Since all the datasets have been validated already, this means we've got an inconsistency between our SQL and the cache manager, which is an internal server error. I think this should be `APIInternalError` with a message saying that the expected dataset ID is missing from the cache manager.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:123]: `tarball.name` should be the same. Also this gets complicated by Riya's new `extract` replacement, which is implicitly adding the `name`. (In this case, the PR that goes in last will need to fix this up.)

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:130]: We already know that all datasets have the same `benchmark` value, so you can save that during initialization and just use it here instead of re-fetching.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/compare_dataset.py:140]: I realized during Riya's review that we should probably be careful that all consumers of the cache manager file streams should `stream_file.close()` after finishing to be sure that we don't accumulate too much IO buffer space before Python gets around to reclaiming dead objects.

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Great start, but I think there are some better patterns you could use.

**Review by dbutenhof — COMMENTED** [review]
Some more comments: partly (sorry) because I was too lazy before to go through the unit tests.

File: lib/pbench/test/unit/server/test_compare_datasets.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_compare_datasets.py:57]: This is returning the (mock) `Tarball` _class_...I think we are supposed to be returning a `Tarball` _instance_ here.  I.e., this should be:
```python
        return Tarball()
```

Ditto line 133.

File: lib/pbench/test/unit/server/test_compare_datasets.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_compare_datasets.py:150]: I can't really explain how (the fact that line 133 is returning a class object instead of an instance might be contributing to it, although the botched mock at line 144 is probably the source...) but this test isn't working.  That is, looking at [the coverage report](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3470/4/cobertura/server_api_resources/compare_dataset_py/#137), the `raise APIInternalError` at [`compare_dataset.py` line 137](https://github.com/distributed-system-analysis/pbench/blob/ceff0e4e54e6b131d4c75941699fae82c2f4d53f/lib/pbench/server/api/resources/compare_dataset.py#L137) is never executed...so the assertion here at line 148 is passing on a _different_ internal error....

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (webbnh) [reply lib/pbench/server/api/resources/compare_dataset.py:99]: Alternately,
```python
        benchmark_choice = Metadata.getvalue(datasets[0], "dataset.metalog.pbench.script")
        for dataset in datasets:
            benchmark = Metadata.getvalue(dataset, "dataset.metalog.pbench.script")

            # Validate if all the selected datasets is of same benchmark
            if benchmark != benchmark_choice:
                raise APIAbort(
                    HTTPStatus.BAD_REQUEST,
                    f"Requested datasets must all use the same benchmark; found references to {benchmark_choice} and {benchmark}.",
                )
```

File: lib/pbench/test/unit/server/test_compare_datasets.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_compare_datasets.py:144]: Line 144 looks suspicious:  does `QuisbyProcessing` have an `extract_data()` function, or are you just adding one?...and leaving its `compare_csv_to_json()` un-mocked....

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (webbnh) [inline lib/pbench/server/api/resources/compare_dataset.py:99]: We need a test case for this failure.

**Review by webbnh — CHANGES_REQUESTED** [review]
The `test_unsuccessful_get_with_incorrect_data()` test is _not_ working, even though it is passing; and, it appears to have a bad mock setup, which probably isn't helping.  There are a few other gaps in the coverage, one in particular which I would like to see closed is the multiple/non-comparable benchmarks error.  And, I would like to see a test for comparing public datasets owned by someone else from a non-authorized user (which I think should work and pass...).  And, there is a bug which also shows up in two of the tests.

Beyond those, I have a number of other thoughts, suggestions, and nits.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/compare_dataset.py:83]: Good point: we shouldn't be doing this.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/compare_dataset.py:99]: True enough: I'd had the thought that we could report the disparate set of benchmarks the user had selected. If there are only two, your version is more or less equivalent, although the code is a bit more convoluted. I'm not entirely sure the extra level of diagnostics are important: but if we're going to show more than one benchmark name I'd prefer to show them all.

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/compare_dataset.py:130]: I dislike using the overly generic `BAD_REQUEST` intensely, it's just that the more I think about it, the less intensely I dislike it compared to all the other options ... 😦

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/compare_dataset.py:135]: It's a wibbly wobbly timey wimey thing ... yeah, right now, for a fairly instantaneous (and therefore technically correct but globally limited) definition of "now", `Tarball.extract` returns a `str`; but I've been staring at Riya's PR which returns a stream, and the temporal dysphoria is real.

In fact, it's worth noting that the current `Tarball.extract` does a `tarfile.open` and then just lets the object go out of scope. Does `tarfile` have a close? Is it a context manager? Should/can `tarfile.extractfile` be used as a context manager, or closed? Hmm...

But that raises the question, if `stream_file` is actually a `dict[str,str]` where each value is the entire contents of a `.csv` file, should we be passing `InputType.STREAM` to Quisby? Or does this code, apparently much like me, think it's living in Riya's PR?

File: lib/pbench/test/unit/server/test_compare_datasets.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_compare_datasets.py:70]: A lot of this depends on how we define the API contract. Are we allowed to add keys without breaking the contract? Sometimes "yes", sometimes "no". I suppose that in this case I'm inclined to agree that it's pointless and unnecessarily restrictive to assert that there are no other keys. And in fact in some other cases I've added ancillary keys like `errors` ... although, were that to happen here, I'd also want to verify that those are or are not present as appropriate to the test case.

Which is all, in a rambling Friday morning way, to say that I can go either way on this one ...

File: lib/pbench/server/api/resources/compare_dataset.py
Comment (webbnh) [reply lib/pbench/server/api/resources/compare_dataset.py]: > if we're going to show more than one benchmark name I'd prefer to show them all.

I'm not averse to doing that, but, as you say, we would have to move the check to the end of the loop to make that work (and, if that's what we want, then a `set` is a very good approach).  I was just taking the structure that Siddardh had laid out and making it work with a simple variable instead of a `set`...because it could.  🙂

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_compare.py:90]: It would be good to match only the message key here, instead of the whole dictionary (like you did at line 121 and elsewhere).

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_compare.py:13]: `Metadata.getvalue()` takes three parameters, the last of which is optional, so the mock really should do the same.

**Review by webbnh — CHANGES_REQUESTED** [review]
There are still a couple of places where a mock is returning a mock `Tarball` class object where I'm pretty sure it should be returning an instance of the class.  And, there are a couple of other similar sorts of issues (e.g., functions which should be marked `@staticmethod`).  These should be fixed before we merge this PR.

That said, the [test coverage](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3470/8/cobertura/server_api_resources/datasets_compare_py/#111) for `lib/pbench/server/api/resources/datasets_compare.py` is 95%, which is excellent -- it's missing jsut _two_ lines -- consider going for 100%  (we just need one more test case...). 🏁 

Also, there are a few opportunities for "DRYing" the test code out.  There are some code blobs which could be refactored and made common, and, better yet, I think there are some opportunities to collapse whole test functions together using Pytest's "parametrization".

And, there are some smaller items for your consideration.

File: lib/pbench/server/api/resources/datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/datasets_compare.py:79]: I liked it better without the keywords on one line, but this works too. 😆

File: lib/pbench/server/api/resources/datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/server/api/resources/datasets_compare.py:86]: Maybe "Selected dataset benchmarks must match: {benchmark_choice} and {benchmark} cannot be compared."??

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_datasets_compare.py:109]: Ah, yes: as Webb pointed out elsewhere, `return MockTarball()` would be correct.

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_datasets_compare.py:166]: `MockTarball()`

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_datasets_compare.py:15]: Nothing wrong here, it's a creative and convenient lie, but I can't help but be amused that two "uperf" datasets are now "hammerdb".  (And ... would the actual `dataset.metalog.pbench.script` value be "hammerdb" or "hammerDB" or something else? I wonder if we even have any... and again that's completely irrelevant since we just need them not to match. 😆 )

**Review by dbutenhof — CHANGES_REQUESTED** [review]
I think we're getting close here; down to relatively small stuff.

File: lib/pbench/server/api/resources/datasets_compare.py
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_compare.py:79]: Yeah...without the keywords, Black didn't have to wrap the line for us....  🥴

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_compare.py:109]: Yes...but it'll probably force him to correct something else....  😁

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_compare.py]: Based on looking at the coverage, I don't think this test is working correctly.  I think we're expecting this test to raise an `APIInternalError` at [line 110](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3470/11/cobertura/server_api_resources/datasets_compare_py/#110), but Cobertura claims that that line is never executed.  Presumably, the exception is being raised at line 102, instead, but I don't know why.  (Is there some reason why user `test` doesn't have access to datasets `uperf_1` and `uperf_2` such that the `mock_find_dataset()` call is raising an exception?)

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_compare.py:58]: With the change to use parametrization, this function is now used in only one place; consider making it an inner function (or even a lamba, although it's unwieldy) inside `test_datasets_with_different_benchmark()`.

**Review by webbnh — CHANGES_REQUESTED** [review]
Oh, this is so close!

Thanks for doing the parametrization -- that's much nicer, I think (although, we did lose some negative assertions in the process, but I think it's worth it).

There are a few rough edges remaining, which we _could_ ignore, but there are more substantial problems with two of the tests which I'd think should be addressed, and that will give you a chance for some additional polishing.

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (siddardh-ra) [reply lib/pbench/test/unit/server/test_datasets_compare.py]: Yeah @webbnh, It was due to this `raise TarballUnpackError`  as you mentioned in the earlier comment, it expects the instance of the `TarballUnpackError` which I missed. After adding this suggestion 
`raise TarballUnpackError(_tarball_path, f"Testing unpack exception for path {_path}")` , It was working as expected :)

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_datasets_compare.py:153]: Not really that important: but, for the record, this unconditional mock means your "if success" asserts below aren't doing anything beyond verifying that `monkeypatch` worked; only the assert in the `else` is actually interesting. But there are worse things than redundant checks. 😆

Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_compare.py:153]: I don't _think_ that's true (although, my track record on thinking today is not good...):  this test is parametrized, and there are ways other than this mock to generate failures....  (And, the cases where `compare_csv_to_json()` generates failures are tested in tests other than this one.)

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_compare.py:79]: I'm pretty sure that all we need this mock to do is to keep the test from failing (e.g., to prevent it from actually looking for a real path within a real tarball); the behavior that this test is trying to drive comes from the `compare_csv_to_json()` mock.  So, as long as this mock does something innocuous, we're good.

File: lib/pbench/test/unit/server/test_datasets_compare.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_compare.py:153]: > this test is parametrized, and there are ways other than this mock to generate failures.... (And, the cases where `compare_csv_to_json()` generates failures are tested in tests other than this one.)

I'm confused, but I don't think it matters. The `if` case I'm talking about means we succeeded on the strength of this mock, because I don't think there's any other path to success given the setup. The failure cases here go to the `else`, and failure cases elsewhere have nothing to do with this mock or this test case since they're isolated.

More importantly, perhaps, I made comments but approved. In fact I think I'll just resolve these comments. Are we ready to merge this? 😆

Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_compare.py:153]: > The `if` case I'm talking about means we succeeded on the strength of this mock, because I don't think there's any other path to success given the setup.

Correct; but, with different parametrization, we won't get to this mock, and so we need to use different asserts in those cases.

> The failure cases here go to the `else`, and failure cases elsewhere have nothing to do with this mock or this test case since they're isolated.

Agreed; but, since this test tests both failure and success cases, we need the `if` statement.  And, in the success case, those assertions are checking that the value returned by the mock are subsequently returned by the API, so I think they are more than just redundant.  (And, I don't think those scenarios are checked anywhere else.)

> Are we ready to merge this? 😆

I'm waiting to see the output of the tests, which are almost done, now.

**Comment by webbnh** [comment]
So, I just checked build #13, and it's still showing two gaps in the coverage.  I wouldn't block the merge on this basis, except that I think that we think that the existing tests are _supposed_ to cover those gaps.  So, I think we need to at least understand why the tests aren't working as expected.  😞

**Review by webbnh — APPROVED** [review]
At last!

**Comment by dbutenhof** [comment]
> At last!

Now assuming Jenkins and GitHub do their jobs, someone can merge soon. Sigh.

(And, for the record, we had a 100% successful run with 100% coverage of the new module, but Jenkins glitched and couldn't talk to GitHub so the result never propagated. We're hoping for a reprise.)

**Pull request merged**
